### PR TITLE
Refactors downloading and fixes download start time.

### PR DESCRIPTION
**Pull request by garrettmoon**
Download start time was being incorrectly computed since the timer
was started when the image was added to the queue instead of when the
image was actually started downloading.

When attempting to address this, I decided to do some much needed cleanup
and refactoring by moving much of the download work to PINRemoteImageDownloadTask

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/61f76c82596ad74f263782127e643b8e/raw/04848cc2b9d61cfef605abd880e1a3c0f76ce164/log.txt)

File: Source/Classes/Categories/PINRemoteImageTask+Subclassing.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//
+//  Created by Garrett Moon on 5/22/17.
+//  Copyright © 2017 Pinterest. All rights reserved.
+//
+
+#import "PINRemoteImageTask.h"
+
+@interface PINRemoteImageTask (Subclassing)
+
+- (nonnull NSMutableDictionary *)l_callbackBlocks;
```
Comment (Adlai-Holler) [inline Source/Classes/Categories/PINRemoteImageTask+Subclassing.h]: Might as well go in with `NS_ASSUME_NONNULL` here, and `nullable PINResume *_Nullable *` or whatever it is for the resume handler.

File: Source/Classes/PINProgressiveImage.m [hunk 2]
```diff
@@ -102,19 +105,56 @@ - (CFTimeInterval)estimatedRemainingTimeThreshold
...
+- (float)l_bytesPerSecond
 {
-    [self.lock lock];
-        CFTimeInterval startTime = _startTime;
-    [self.lock unlock];
-    return startTime;
+    NSAssert(_dataTask.PIN_startTime != 0, @"Start time needs to be set by now.");
+    CFTimeInterval endTime = CACurrentMediaTime();
+    if (_dataTask.PIN_endTime != 0) {
+        endTime = _dataTask.PIN_endTime;
```
Comment (Adlai-Holler) [inline Source/Classes/PINProgressiveImage.m]: The logic here is a little bit hard to follow. Let's do something like `CFTimeInterval endTime = (_dataTask.PIN_endTime ?: CACurrentMediaTime())`

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/ae4d5467f38c721fd8f862f505347116/raw/019ed558b8f4998517fd7e906b9f4ec18b23347e/log.txt)

File: Source/Classes/PINRemoteImageManager.m [hunk 3]
```diff
@@ -1062,26 +1018,14 @@ - (void)cancelTaskWithUUID:(nonnull NSUUID *)UUID storeResumeData:(BOOL)storeRes
                 [strongSelf.canceledTasks addObject:UUID];
             }
             
-            if ([taskToEvaluate cancelWithUUID:UUID manager:strongSelf]) {
+            if ([taskToEvaluate cancelWithUUID:UUID resume:storeResumeData ? &resume : nil]) {
```
Comment (Adlai-Holler) [inline Source/Classes/PINRemoteImageManager.m]: Silly nit: this should be `NULL` since its type is `id *`

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/a4195f196880db64673e932f6d9fc707/raw/fc817048981e3a2618394acced02bf42f23d4c84/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/28ac2ecf87b06dd81cea3df5f7f90a31/raw/a9ab4a6ca4779a26a968484cddbf32b0b3a4b139/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/e44dea60764812e2bc18257e8229c038/raw/a8ca6220a6542a58edac146b59b7eb3bca7a05a6/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
